internal/impl/kafka/enterprise/redpanda_migrator_offsets_output.go

```diff
@@ -265,6 +317,36 @@ func (w *redpandaMigratorOffsetsWriter) Write(ctx context.Context, msg *service.
 					delete(partitionOffsets, partition)
 				}
 				offset.Metadata = offsetMetadata
+
+				// As an optimisation to try and avoid unnecessary duplicates in the common case, we check if the
+				// received offset update was triggered when a consumer read the last record of the source topic. In
+				// this special case, we check if the matching offset in the destination topic (returned by
+				// `ListOffsetsAfterMilli`) also points to the end record. If it does, then we can fetch the current end
+				// offset of the destination topic and set the destination consumer offset to that value.
+				// Note: We have to be conservative here and assume there might be duplicates in the destination topic,
+				// so we can't just update the offset to the end offset without first checking if
+				// `ListOffsetsAfterMilli` actually returned the offset of the end record.
+				if isEndOffset {
+					lastOffset, err := w.getLastRecordOffset(ctx, topic, partition)
+					if err != nil {
+						return err
+					}
+
+					if offset.At == lastOffset {
+						offsets, err := w.client.ListEndOffsets(ctx, topic)
+						if err != nil {
+							return fmt.Errorf("failed to read the end offset for topic %q and partition %q: %s", topic, partition, err)
+						}
+
+						committedOffset, ok := offsets.Lookup(topic, partition)
+						if !ok {
+							return fmt.Errorf("failed to find the end offset for topic %q and partition %q: %s", topic, partition, err)
+						}
+
+						offset.At = committedOffset.Offset
```

Comment:
This is almost the same as doing `offset.At++` but is racy right? If new records come in during `getLastRecordOffset` and `ListEndOffsets` then we can skip records because ListEndOffsets will pick up the new records?

I am also trying to think through if comparing the offsetCommitTimestamp to the record timestamp from `getLastRecord` would also be benifical to ensure we're not skipping data... Let me noodle on an algorithm here.

Reply:
> This is almost the same as doing offset.At++ but is racy right?

Yep! I should've called that `committedOffset` variable `endOffset`. Fixing now...

It's racy indeed either way... I just realised now that I could get new messages written to the topic between `getLastRecordOffset` and `ListEndOffsets` which will lead to a fast forward... I was thinking initially that I'd have to call `getLastRecordOffset` again after `ListEndOffsets` to check if `lastOffset` changed in the meantime. Not sure why I changed my mind 🤦‍♂️ 

> Let me noodle on an algorithm here.

OK, curious if we can get away with simpler logic. In practice `isEndOffset` will be `true` frequently only when topics don't get a lot of traffic and consumers catch up immediately, so maybe the performance isn't a concern.